### ISSUE TITLE
autocast + torch.no_grad inference cause backward graph nodes to be lost

### ISSUE THREAD

**Issue by yupeifengyannis** [pytorch/pytorch#105211]
### 🐛 Describe the bug

In autocast, after the inference operation of torch.no_grad, I ran the forward and backward of the network,  found that the gradient of some parameters has not been calculated. This question is from the real scene of bevformer's amp training 
https://github.com/fundamentalvision/BEVFormer/blob/master/projects/mmdet3d_plugin/bevformer/detectors/bevformer.py#L158。 
```
import torch
import torchvision
device = 'cuda'
net = torchvision.models.resnet18().to(device)
input = torch.randn(1,3,224,224).to(device)

with torch.cuda.amp.autocast(True):
    net.eval()
    with torch.no_grad():
        _ = net(input)
        net.train()
    #torch.clear_autocast_cache()
    out = net(input)

grad = torch.randn_like(out.data)
out.backward(grad)

for name, param in net.named_parameters():
    if param.grad != None:
        print(f"{name} parameter's grad is not None")
    else:
        print(f"{name} parameter's grad is None")

```

### Versions

Collecting environment information...
PyTorch version: 1.9.0
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.10

Python version: 3.7.10 (default, Feb 26 2021, 18:47:35)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-4.4.0-186-generic-x86_64-with-debian-buster-sid
Is CUDA available: True
CUDA runtime version: 10.2.89
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: Tesla V100S-PCIE-32GB
GPU 1: Tesla V100S-PCIE-32GB
GPU 2: Tesla V100S-PCIE-32GB
GPU 3: Tesla V100S-PCIE-32GB

Nvidia driver version: 525.85.12
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              96
On-line CPU(s) list: 0-95
Thread(s) per core:  2
Core(s) per socket:  24
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Gold 6240R CPU @ 2.40GHz
Stepping:            7
CPU MHz:             3199.968
CPU max MHz:         4000.0000
CPU min MHz:         1000.0000
BogoMIPS:            4799.99
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            36608K
NUMA node0 CPU(s):   0-23,48-71
NUMA node1 CPU(s):   24-47,72-95
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb invpcid_single intel_pt ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx avx512f rdseed adx smap clflushopt clwb avx512cd xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req pku md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] flake8==5.0.4
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.20.2
[pip3] torch==1.9.0
[pip3] torchelastic==0.2.0
[pip3] torchtext==0.10.0
[pip3] torchvision==0.10.0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              h6bb024c_0    nvidia
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.2.0           h06a4308_296
[conda] mkl-service               2.3.0            py37h27cfd23_1
[conda] mkl_fft                   1.3.0            py37h42c9631_2
[conda] mkl_random                1.2.1            py37ha9443f7_2
[conda] numpy                     1.20.2           py37h2d18471_0
[conda] numpy-base                1.20.2           py37hfae3a4d_0
[conda] pytorch                   1.9.0           py3.7_cuda10.2_cudnn7.6.5_0    pytorch
[conda] torch                     1.9.0a0+git574ab98          pypi_0    pypi
[conda] torchelastic              0.2.0                    pypi_0    pypi
[conda] torchtext                 0.10.0                     py37    pytorch
[conda] torchvision               0.10.0               py37_cu102    pytorch

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by albanD** [comment]
Also you should not mix eval and train in a single autocast context. In general, you should not even have multiple iterations.

**Comment by haydn-jones** [comment]
I'm seeing something similar, though I'm not playing with `train()`/`eval()` at all. Is this expected behavior?

```Python
import torch
import torch.nn as nn
import torch.nn.functional as F

l1 = nn.Linear(2, 2).cuda()
l2 = nn.Linear(2, 2).cuda()
l2.load_state_dict(l1.state_dict())

x = torch.randn(2, 2).cuda()
y = torch.randn(2, 2).cuda()

##################################
# Here l1_loss does not have a grad_fn
##################################
with torch.autocast(device_type='cuda', dtype=torch.float16):
    with torch.no_grad():
        xhat = l1(x)
    yhat = l1(xhat)
    l1_loss = F.mse_loss(yhat, y)
l1_loss.backward()

##################################
# Here l2_loss DOES have a grad_fn
##################################
with torch.autocast(device_type='cuda', dtype=torch.float16):
    xhat = l2(x).detach()
    yhat = l2(xhat)
    l2_loss = F.mse_loss(yhat, y)
l2_loss.backward()
```

**Comment by ezyang** [comment]
Hmm, I actually think your example is unexpected, can you file a separate bug for it.

**Comment by haydn-jones** [comment]
Yes, will do.
